blob/blob.go

```diff
@@ -0,0 +1,185 @@
+}
+
+type Table struct {
+	Name string
+	drv  *Driver
+	c    *http.Client
+}
+
+func New(crate_url string) (*Driver, error) {
+	db, err := sql.Open("crate", crate_url)
+	if err != nil {
+		return nil, err
+	}
+	return &Driver{
+		url: crate_url,
+		db:  db,
+	}, nil
+}
+
+// NewTable create new blob table with name and extra int to specify shards and replicas
+func (d *Driver) NewTable(name string, shards ...int) (*Table, error) {
+	sql := fmt.Sprintf(
+		"create blob table %s",
+		name,
+	)
+	if len(shards) == 1 {
+		sql = fmt.Sprintf(
+			"create blob table %s clustered into %d shards",
+			name, shards[0],
+		)
+	}
+	if len(shards) >= 2 {
+		sql = fmt.Sprintf(
+			"create blob table %s clustered into %d shards with (number_of_replicas=%d)",
+			name, shards[0], shards[1],
+		)
+	}
+	_, err := d.db.Exec(sql)
+	if err != nil {
+		return nil, err
+	}
+	return &Table{
+		Name: name,
+		drv:  d,
+		c:    new(http.Client),
+	}, nil
+}
+
+func (d *Driver) GetTable(name string) (*Table, error) {
+	_, err := d.db.Exec(
+		"select count(*) from information_schema.tables where table_name = '?' and schema_name = 'blob'",
+		name,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &Table{
+		Name: name,
+		drv:  d,
+		c:    new(http.Client),
+	}, nil
+}
+
+type Record struct {
+	Digest       string
+	LastModified time.Time
+}
+
+func Sha1Digest(r io.Reader) string {
+	h := sha1.New()
+	io.Copy(h, r)
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func (t *Table) Upload(digest string, r io.Reader) (*Record, error) {
+	url := fmt.Sprintf("%s/_blobs/%s/%s", t.drv.url, t.Name, digest)
+	req, err := http.NewRequest("PUT", url, r)
+	if err != nil {
+		return nil, err
+	}
+	_, err = t.c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return &Record{
+		Digest: digest,
+	}, nil
+}
+
+// UploadEx upload a io.ReadSeeker directly
+func (t *Table) UploadEx(r io.ReadSeeker) (*Record, error) {
+	digest := Sha1Digest(r)
+	url := fmt.Sprintf("%s/_blobs/%s/%s", t.drv.url, t.Name, digest)
+	_, err := r.Seek(0, 0)
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest("PUT", url, r)
+	if err != nil {
+		return nil, err
+	}
+	_, err = t.c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return &Record{
+		Digest: digest,
+	}, nil
+}
+
+func (t *Table) List() (*sql.Rows, error) {
+	query := fmt.Sprintf("select digest, last_modified from blob.%s", t.Name)
+	rows, err := t.drv.db.Query(query)
+	if err != nil {
+		return nil, err
+	}
+	return rows, err
+}
+
+func (t *Table) Has(digest string) (bool, error) {
+	url := fmt.Sprintf("%s/_blobs/%s/%s", t.drv.url, t.Name, digest)
+	req, err := http.NewRequest("HEAD", url, nil)
+	if err != nil {
+		return false, err
+	}
+	resp, err := t.c.Do(req)
+	if err != nil {
+		return false, err
+	}
+	if resp.StatusCode == http.StatusOK {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (t *Table) Download(digest string) (io.ReadCloser, error) {
+	url := fmt.Sprintf("%s/_blobs/%s/%s", t.drv.url, t.Name, digest)

```

Comment:
perhaps the `Table` struct could have a method like
```go
func (t *Table) url() string {
  return fmt.Sprintf("%s/_blobs/%s/%s", t.drv.url, t.Name, digest)
}
```

since I saw that you will need that url at each method